### PR TITLE
RFC: Document data/network freeze policies

### PR DESCRIPTION
**Pull request by bdarnell**
It's almost time to stop making backwards-incompatible changes.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4499)

### CODE REVIEW

**Comment by tschottdorf** [comment]
I have no practical experience making migrations seamless (only as a user, but seamless they never were). The text LGTM as a starting point. What are you proposing in this RFC? Making an actual schedule? Or just having those those stages (in other words, what's the pending/completed state for this RFC)?

File: docs/RFCS/freeze.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+will be conservative and cross-version communication will be limited.
+That makes these version numbers a blunt instrument to be reserved for
+major changes.
+
+* The on-disk format (via a file that lives outside RocksDB). Servers
+  will refuse to load a database with a higher version number than
+  they understand.
+* The network protocol (via GRPC header). Servers and clients will
+  treat a higher version number than they understand as an error.
+* Gossip (perhaps via a new node attribute). The rebalance/allocation
```
Comment (spencerkimball) [inline docs/RFCS/freeze.md]: Why not be explicit here and say we'll add the version to the `NodeDescriptor` proto.

**Comment by bdarnell** [comment]
The point of this RFC is mainly to have a place to point to to discuss the policy. It's a mix of A) last call for breaking changes before the freeze and B) documentation of what stage we're currently in. A schedule would be ideal.

File: docs/RFCS/freeze.md (hunk 1 above)
Comment (bdarnell) [inline docs/RFCS/freeze.md]: The allocator currently only looks at attributes, not a full NodeDescriptor. I haven't decided yet whether it makes more sense to treat the version as an attribute or introduce the rest of the NodeDescriptor into the allocation process.

File: docs/RFCS/freeze.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+To facilitate future changes, we will introduce version numbers at
+several levels. Initially the behavior around these version numbers
+will be conservative and cross-version communication will be limited.
+That makes these version numbers a blunt instrument to be reserved for
+major changes.
+
+* The on-disk format (via a file that lives outside RocksDB). Servers
+  will refuse to load a database with a higher version number than
+  they understand.
+* The network protocol (via GRPC header). Servers and clients will
```
Comment (spencerkimball) [inline docs/RFCS/freeze.md]: I think this is starting with too blunt a mechanism, as it will defeat all non-upgraded clients once a server is upgraded. I propose we add the following nuance to network protocol:

RPC request headers would specify:

```
  int client_version;
  int required_server_version;
```

There would be constants in the server code:
 ```
const(
  ServerVersion = X
  MinSupportedVersion = Y
)

RPCs specifying a higher `required_server_version` than `ServerVersion` would fail. Similarly, clients specifying a lower `required_server_version` than `MinSupportedVersion` would fail. Otherwise, the server would migrate incoming RPCs from `client_version` to `ServerVersion` and then migrate outgoing RPC responses from `ServerVersion` to `client_version`.

I realize this is extra complication. However, putting it in place now when we have zero migrations to implement will set the stage for the first real migration, without requiring any real effort. We used this system successfully at Viewfinder and I was never dissatisfied with it.

File: docs/RFCS/freeze.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+To facilitate future changes, we will introduce version numbers at
+several levels. Initially the behavior around these version numbers
+will be conservative and cross-version communication will be limited.
+That makes these version numbers a blunt instrument to be reserved for
+major changes.
+
+* The on-disk format (via a file that lives outside RocksDB). Servers
+  will refuse to load a database with a higher version number than
+  they understand.
```
Comment (petermattis) [inline docs/RFCS/freeze.md]: We have on-disk formats at several levels. There is the low-level encoding of MVCC keys. There are the `util/encoding` routines. There are the structure of various protobufs that are stored on disk. And there is the mapping of SQL data to keys and values. 

Some of these formats are easier to change than others. For example, with a bit of care new fields can be added to protobufs. Changing the MVCC key-encoding could be done by draining a server of data, wiping anything that remains and then bringing a new server up with the new data.

Changing the SQL to KV mapping is more problematic because the data format is spread across nodes. My initial thought is that we don't try to upgrade the format for a table. That is, the table data format is immutable for the life time of the table. What we might do, though, is support multiple table data formats. For example, implementing interleaved tables might require bumping the table data format version and all binaries which didn't support that version would be able to politely decline to operate on such a table.

File: docs/RFCS/freeze.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+# Summary
+
+This RFC outlines the plan for freezing our data formats and network
+protocols.
+
+# Motivation
+
+We currently make backwards-incompatible changes to data formats
+without providing any means to upgrade without data loss. This will
+need to stop before beta for obvious reasons.
```
Comment (mjibson) [inline docs/RFCS/freeze.md]: Have we considered not making this guarantee for the beta release? This would allow us to, for example, provide a dump/restore tool instead of being forced to write migration logic for all migrations ever. This could be useful if one of the more difficult-to-change data formats needs a revision. It's obvious that we need this guarantee as we near a stable release, but we may be biting off a large support burden here for a relatively small number of users who may not be using it seriously.

File: docs/RFCS/freeze.md (hunk 2 above)
Comment (bdarnell) [inline docs/RFCS/freeze.md]: This worked at viewfinder because the centralized server provided a one-way guarantee: the server had to support old clients but the clients never had to support an old server. When the server may be older than the client, the client must choose one of two options:
* Pessimistically speak an older protocol until it has learned the server's version
* Optimistically speak the newest protocol until it gets a "version not supported" error from the server. 

Either way, this can be achieved whether each node advertises a single version or a range of versions; I don't think making this into a range adds much.

Also note that this version number is meant to cover major changes comparable to our move from net/rpc to our homegrown codec to grpc. We wouldn't use this mechanism for the kinds of changes that we made with viewfinder's migration system. Those would be done by adding an in-band version number to the affected request proto (or maybe just `BatchRequest`) and a viewfinder-like migration scheme (when the change is not trivialized by protobuf backwards-compatibility).

File: docs/RFCS/freeze.md (hunk 3 above)
Comment (bdarnell) [inline docs/RFCS/freeze.md]: Are you suggesting that we introduce a format version number at the table (TableDescriptor? Replica?) level?

File: docs/RFCS/freeze.md (hunk 4 above)
Comment (bdarnell) [inline docs/RFCS/freeze.md]: I considered adding a "stage 1.5" in which migration would be possible but would have to go through some sort of dump/restore tool. I dropped it because I thought that anything the dump/restore tool could do could also be done at store startup (although on further reflection there are some things that would probably be easier with this process).

Is it worth providing this guarantee in beta? That's one of the reasons why this is an RFC. My opinion is that the guarantees discussed here are an essential part of the difference between alpha and beta. We need our beta users to start putting cockroachdb into production-like scenarios, and delaying the point at which we reach data-format stability delays the point at which we can get useful feedback.

**Comment by JackKrupansky** [comment]
Whatever difficulties alpha/beta/1.0 storage format transitions incur will most likely be relatively small compared to format transitions down the road, like 1.0 to 2.0, 4.0 to 5.0, 5.0 to 10.0 as the level of features, granularity of types, and desire (and pressure) for optimizations only increases. Not to mention the issue of rolling upgrades where portions of a cluster will still be on an older release while portions incrementally shif to a new release.

Not to mention the fact that RocksDB itself will likely evolve over time.

In short, incremental, online, real-time migration is a fact of life these days, so best to get the infrastructure for it plumbed into the code ASAP.

File: docs/RFCS/freeze.md (hunk 3 above)
Comment (petermattis) [inline docs/RFCS/freeze.md]: I don't think we have to do anything at this time, because for protobufs like `TableDescriptor` we can add an optional version field when we support another version. What I was proposing is that we enumerate some of the major areas which are sensitive to versioning. We have the mvcc key-encoding, the specific way we're using RocksDB (single column family), the key encoding routines and the mapping of SQL table data to KV data.

Comment (johnmuth81) [inline docs/RFCS/freeze.md]: > On Feb 21, 2016, at 5:11 PM, Peter Mattis <notifications@github.com> wrote:
> 
> In docs/RFCS/freeze.md <https://github.com/cockroachdb/cockroach/pull/4499#discussion_r53578683>:
> 
> > +It is difficult to come up with a universal migration strategy, since
> > +different changes will require different approaches (for example,
> > +`.proto` changes could perhaps be made by rewriting data on disk at
> > +startup, while changes to key construction may require the change to
> > +be coordinated in a distributed fashion). Therefore we leave the
> > +specifics of a migration process until the need arises.
> > +
> > +To facilitate future changes, we will introduce version numbers at
> > +several levels. Initially the behavior around these version numbers
> > +will be conservative and cross-version communication will be limited.
> > +That makes these version numbers a blunt instrument to be reserved for
> > +major changes.
> > +
> > +* The on-disk format (via a file that lives outside RocksDB). Servers
> > +  will refuse to load a database with a higher version number than
> > +  they understand.
> I don't think we have to do anything at this time, because for protobufs like TableDescriptor we can add an optional version field when we support another version. What I was proposing is that we enumerate some of the major areas which are sensitive to versioning. We have the mvcc key-encoding, the specific way we're using RocksDB (single column family), the key encoding routines and the mapping of SQL table data to KV data.
> 
I'm not sure that’s true. 

As I read it, this is a requirement that old software refuse to work with newer formats. For this to work, the old software must know about a version field in order to check it. 

If we add it as an optional field in the future, then existing software will ignore it.

Regards,
John Muth

Comment (petermattis) [inline docs/RFCS/freeze.md]: That's a good point, John. Guess we should add a `TableDescriptor.Version` field now.

**Comment by bdarnell** [comment]
I've updated the doc with some feedback from this thread, and a proposed date. It's also time to set a schedule. I propose March 24 as the last day for backwards-incompatible changes without a migration path. We've had some incompatible changes recently; are there any more in the pipeline?

**Comment by JackKrupansky** [comment]
Is the version info per-store (RocksDB instance) or global for the node/CockroachDB instance. As per the proposal, "The on-disk format (via a file that lives outside RocksDB). Servers will refuse to load a database..."

Is "database" being used in the Rocks sense, so that each store on a node is a separate database, or in the CockroachDB sense of all tables on the node (which is technically only a portion of a distributed database)? Hmmm... I suspect it could be a little of both - technically CockroachDB could ignore an individual store which is not of the proper version.

This may only be an issue when it comes to upgrading nodes when modifying the KV structure may be required - presumably that would happen for each store, and it would be possible that one/some of the stores on a node have been successfully upgraded but not all of them at some intermediate stage.

File: docs/RFCS/freeze.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+In stage 1, we require backwards-compatibility with data written by
+any previous stage 1 build. It should always be possible to upgrade by
+stopping all of the old nodes and then bringing up the new version.
+It's OK at this stage if old and new versions cannot be run
+concurrently, or if the migration process takes some time (e.g.
+rewriting all data before the node can start up).
+
+It is acceptable at this stage if the process is somewhat manual (e.g.
+stop all the nodes, run `cockroach migrate` on each one, then restart
```
Comment (petermattis) [inline docs/RFCS/freeze.md]: I agree with the acceptability of a manual process. Rather than a per-node `migrate` command, I think we should initially target a `dump/restore` process that operates at the level of SQL metadata (tables, indexes, permissions, etc) and data (rows). I think this approach would handle more cases (a per-node `migrate` seems like it would have difficulty with a SQL data format change) and has a more straightforward story for downgrades. It definitely isn't as operator friendly, but I bet we'll want the flexibility.

Comment (bdarnell) [inline docs/RFCS/freeze.md]: Yeah, a global dump/restore is probably a better example, since anything that could be run on one node in isolation could be run automatically at startup. I'll update the text, but anyway this is just an example and we'll decide the real process on a per-migration basis.

**Comment by bdarnell** [comment]
@JackKrupansky All the version numbers discussed here are per-node; there is no notion of a cluster-wide version because upgrades necessarily happen asynchronously (except in the stop-the-world scenario permitted during beta). The closest thing we have to a cluster-wide version is that gossiped version numbers could be used to coordinate a cluster-wide switchover once a majority (or supermajority) of nodes have been upgraded.
